File: specification/compute/resource-manager/Microsoft.Compute/Compute/stable/2026-04-01/examples-java/virtualMachineExamples/VirtualMachine_Create_WithAdditionalDiskProperties.java

import com.azure.resourcemanager.compute.fluent.models.VirtualMachineInner;
import com.azure.resourcemanager.compute.models.AdditionalDiskProperties;
import com.azure.resourcemanager.compute.models.CachingTypes;
import com.azure.resourcemanager.compute.models.DataDisk;
import com.azure.resourcemanager.compute.models.DiskApiVersion;
import com.azure.resourcemanager.compute.models.DiskCreateOptionTypes;
import com.azure.resourcemanager.compute.models.HardwareProfile;
import com.azure.resourcemanager.compute.models.ImageReference;
import com.azure.resourcemanager.compute.models.ManagedDiskParameters;
import com.azure.resourcemanager.compute.models.NetworkInterfaceReference;
import com.azure.resourcemanager.compute.models.NetworkProfile;
import com.azure.resourcemanager.compute.models.OSDisk;
import com.azure.resourcemanager.compute.models.OSProfile;
import com.azure.resourcemanager.compute.models.StorageAccountTypes;
import com.azure.resourcemanager.compute.models.StorageProfile;
import com.azure.resourcemanager.compute.models.VirtualMachineDiskNetworkAccessPolicy;
import com.azure.resourcemanager.compute.models.VirtualMachineDiskProperties;
import com.azure.resourcemanager.compute.models.VirtualMachineSizeTypes;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * Samples for VirtualMachines CreateOrUpdate.
 */
public final class Main {
    /*
     * x-ms-original-file: 2026-04-01/virtualMachineExamples/VirtualMachine_Create_WithAdditionalDiskProperties.json
     */
    /**
     * Sample code: Create a vm with additional disk properties (network access policy and tier).
     * 
     * @param manager Entry point to ComputeManager.
     */
    public static void createAVmWithAdditionalDiskPropertiesNetworkAccessPolicyAndTier(
        com.azure.resourcemanager.compute.ComputeManager manager) {
        manager.serviceClient().getVirtualMachines().createOrUpdate("myResourceGroup", "myVM", new VirtualMachineInner()
            .withLocation("westus")
            .withHardwareProfile(new HardwareProfile().withVmSize(VirtualMachineSizeTypes.STANDARD_D4S_V3))
            .withStorageProfile(new StorageProfile()
                .withImageReference(new ImageReference().withPublisher("MicrosoftWindowsServer")
                    .withOffer("WindowsServer").withSku("2022-datacenter-azure-edition").withVersion("latest"))
                .withOsDisk(new OSDisk().withName("myVMosdisk").withCaching(CachingTypes.READ_WRITE)
                    .withCreateOption(DiskCreateOptionTypes.FROM_IMAGE)
                    .withManagedDisk(new ManagedDiskParameters().withStorageAccountType(StorageAccountTypes.PREMIUM_LRS)
                        .withAdditionalDiskProperties(
                            new AdditionalDiskProperties().withManagedDiskProperties(new VirtualMachineDiskProperties()
                                .withNetworkAccessPolicy(VirtualMachineDiskNetworkAccessPolicy.ALLOW_PRIVATE)
                                .withDiskAccessId(
                                    "/subscriptions/{subscription-id}/resourceGroups/myResourceGroup/providers/Microsoft.Compute/diskAccesses/myDiskAccess")))))
                .withDataDisks(Arrays.asList(new DataDisk().withLun(0).withName("myDataDisk")
                    .withCreateOption(DiskCreateOptionTypes.EMPTY).withDiskSizeGB(1024)
                    .withManagedDisk(new ManagedDiskParameters().withStorageAccountType(StorageAccountTypes.PREMIUM_LRS)
                        .withAdditionalDiskProperties(new AdditionalDiskProperties()
                            .withManagedDiskProperties(new VirtualMachineDiskProperties().withTier("P30")
                                .withNetworkAccessPolicy(VirtualMachineDiskNetworkAccessPolicy.ALLOW_PRIVATE)
                                .withDiskAccessId(
                                    "/subscriptions/{subscription-id}/resourceGroups/myResourceGroup/providers/Microsoft.Compute/diskAccesses/myDiskAccess"))))))
                .withDiskApiVersion(DiskApiVersion.TWO_ZERO_TWO_SIX_ZERO_THREE_ZERO_TWO))
            .withOsProfile(new OSProfile().withComputerName("myVM").withAdminUsername("{your-username}")
                .withAdminPassword("fakeTokenPlaceholder"))
            .withNetworkProfile(new NetworkProfile().withNetworkInterfaces(Arrays.asList(new NetworkInterfaceReference()
                .withId(
                    "/subscriptions/{subscription-id}/resourceGroups/myResourceGroup/providers/Microsoft.Network/networkInterfaces/{existing-nic-name}")
                .withPrimary(true)))),
            null, null, com.azure.core.util.Context.NONE);
    }

    // Use "Map.of" if available
    @SuppressWarnings("unchecked")
    private static <T> Map<String, T> mapOf(Object... inputs) {
        Map<String, T> map = new HashMap<>();
        for (int i = 0; i < inputs.length; i += 2) {
            String key = (String) inputs[i];
            T value = (T) inputs[i + 1];
            map.put(key, value);
        }
        return map;
    }
}

File: specification/compute/resource-manager/Microsoft.Compute/Compute/stable/2026-04-01/examples-js/virtualMachineExamples/VirtualMachine_Create_WithAdditionalDiskProperties.js
const { ComputeManagementClient } = require("@azure/arm-compute");
const { DefaultAzureCredential } = require("@azure/identity");

/**
 * This sample demonstrates how to the operation to create or update a virtual machine. Please note some properties can be set only during virtual machine creation.
 *
 * @summary the operation to create or update a virtual machine. Please note some properties can be set only during virtual machine creation.
 * x-ms-original-file: 2026-04-01/virtualMachineExamples/VirtualMachine_Create_WithAdditionalDiskProperties.json
 */
async function createAVmWithAdditionalDiskPropertiesNetworkAccessPolicyAndTier() {
  const credential = new DefaultAzureCredential();
  const subscriptionId = "{subscription-id}";
  const client = new ComputeManagementClient(credential, subscriptionId);
  const result = await client.virtualMachines.createOrUpdate("myResourceGroup", "myVM", {
    location: "westus",
    hardwareProfile: { vmSize: "Standard_D4s_v3" },
    storageProfile: {
      imageReference: {
        sku: "2022-datacenter-azure-edition",
        publisher: "MicrosoftWindowsServer",
        version: "latest",
        offer: "WindowsServer",
      },
      diskApiVersion: "2026-03-02",
      osDisk: {
        caching: "ReadWrite",
        managedDisk: {
          storageAccountType: "Premium_LRS",
          additionalDiskProperties: {
            managedDiskProperties: {
              networkAccessPolicy: "AllowPrivate",
              diskAccessId:
                "/subscriptions/{subscription-id}/resourceGroups/myResourceGroup/providers/Microsoft.Compute/diskAccesses/myDiskAccess",
            },
          },
        },
        name: "myVMosdisk",
        createOption: "FromImage",
      },
      dataDisks: [
        {
          lun: 0,
          name: "myDataDisk",
          createOption: "Empty",
          diskSizeGB: 1024,
          managedDisk: {
            storageAccountType: "Premium_LRS",
            additionalDiskProperties: {
              managedDiskProperties: {
                networkAccessPolicy: "AllowPrivate",
                diskAccessId:
                  "/subscriptions/{subscription-id}/resourceGroups/myResourceGroup/providers/Microsoft.Compute/diskAccesses/myDiskAccess",
                tier: "P30",
              },
            },
          },
        },
      ],
    },
    osProfile: {
      adminUsername: "{your-username}",
      computerName: "myVM",
      adminPassword: "{your-password}",
    },
    networkProfile: {
      networkInterfaces: [
        {
          id: "/subscriptions/{subscription-id}/resourceGroups/myResourceGroup/providers/Microsoft.Network/networkInterfaces/{existing-nic-name}",
          primary: true,
        },
      ],
    },
  });
  console.log(result);
}
